File: Top Interview Questions/Array/Move Zeroes.swift
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/567/

class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        var numberOfZeroes = 0
        var i = 0
        while (i < nums.count - numberOfZeroes) {
            if nums[i] == 0 {
                nums.remove(at: i)
                nums.append(0)
                numberOfZeroes += 1
            } else {
                i += 1
            }
        }
    }
}

class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        if nums.count == 1 { return }
        
        var slow: Int?
        
        for fast in 0..<nums.count {
            if nums[fast] == 0 {
                continue
            }
            
            if slow == nil {
                if fast == 0 {
                    slow = 0
                    continue
                }
                nums[0] = nums[fast]
                nums[fast] = 0
                slow = 0
                continue
            }
            
            if fast - slow! > 0 && nums[slow! + 1] == 0 {
                nums[slow! + 1] = nums[fast]
                nums[fast] = 0
            }
            slow! += 1
        }
    }
}

File: Array/Move Zeroes.go
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/567/

func moveZeroes(nums []int)  {
    nums = moveZero(nums)
}

func moveZero(nums []int) []int {
    for index, value := range nums {
        if value != 0 {
            continue
        }
        firstPart := nums[:index]
        secondPart := nums[index + 1:]
        secondPartMoved := moveZero(secondPart)
        full := append(firstPart, secondPartMoved...)
        return append(full, 0)
    }
    return nums
}
